File: JavaCheatSheet/src/main/java/com/github/jameswdelancey/javaCheatSheet/Part13.java
package com.github.jameswdelancey.javaCheatSheet;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

class Part13 {
    static class Node {
        String value;
        List<Node> edges;

        Node(String value) {
            this.value = value;
            this.edges = new ArrayList<>();
        }

        void display() {
            System.out.println("Value: " + value);
            for (Node edge : edges) {
                edge.display();
            }
            System.out.println("UP");
        }
    }

    public static void main(String[] args) {
        Node e = new Node("E");
        Node d = new Node("D");
        d.edges.add(e);
        Node b = new Node("B");
        Node c = new Node("C");
        c.edges.add(d);
        c.edges.add(e);
        Node a = new Node("A");
        a.edges.add(b);
        a.edges.add(c);
        a.edges.add(d);
        a.edges.add(e);
        WeakReference<Node> dWeak = new WeakReference<>(d);
        a.display();
        a = null; // Drop a
        d = null; // Drop d
        System.gc(); // Explicitly trigger garbage collection
        Node dStrong = dWeak.get();
        if (dStrong != null) {
            System.out.println("D is still alive: " + dStrong.value);
            dStrong.display();
        } else {
            System.out.println("D is dead");
        }
        System.out.println("Done");
    }
}

class Part13_2 {

    static class Cat {
        private int timesSpoken = 0;

        void report() {
            System.out.println("Spoken " + timesSpoken + " times");
        }

        void speak() {
            timesSpoken++;
            System.out.println("Meow");
        }
    }

    public static void main(String[] args) {
        Cat cat = new Cat();
        cat.speak();
        cat.speak();
        cat.speak();
        cat.report();
    }
}

File: rustCheatSheet/src/part13.rs
use std::rc::Rc;

struct Node {
    value: &'static str,
    edges: Vec<Rc<Node>>,
}

impl Node {
    fn display(&self) {
        println!("Value: {}", self.value);
        for edge in &self.edges {
            edge.display();
        }
        println!("UP");
    }
}
impl Drop for Node {
    fn drop(&mut self) {
        println!("Dropping {}", self.value);
    }
}

#[test]
fn main3() {
    let e = Rc::new(Node {
        value: "E",
        edges: vec![],
    });
    let d = Rc::new(Node {
        value: "D",
        edges: vec![e.clone()],
    });
    let a = Node {
        value: "A",
        edges: vec![
            Rc::new(Node {
                value: "B",
                edges: vec![],
            }),
            Rc::new(Node {
                value: "C",
                edges: vec![d.clone(), e.clone()],
            }),
            d.clone(),
            e,
        ],
    };
    let dWeak = Rc::downgrade(&d);
    a.display();
    drop(a);
    drop(d);
    if let Some(dStrong) = dWeak.upgrade() {
        println!("D is still alive: {}", dStrong.value);
        dStrong.display();
    } else {
        println!("D is dead");
    }
    println!("Done");
}

use std::cell::RefCell;
struct Cat2 {
    timesSpoken: RefCell<usize>,
}
impl Cat2 {
    fn report(&self) {
        let timesSpoken = self.timesSpoken.borrow();
        println!("Spoken {} times", timesSpoken);
    }
}
trait Animal {
    fn speak(&self);
}
impl Animal for Cat2 {
    fn speak(&self) {
        let mut timesSpoken = self.timesSpoken.borrow_mut();
        *timesSpoken += 1;
        println!("Meow");
    }
}
#[test]
fn main5() {
    let cat = Cat2 {
        timesSpoken: RefCell::new(0),
    };
    cat.speak();
    cat.speak();
    cat.speak();
    cat.report();
}
